### PR TITLE
ETH JSON RPC Conformance Tests

### PR DESCRIPTION
**Pull request by praetoriansentry**
# Description

We build and work with a wide variety of EVM implementations. We need quick ways to verify adherence to common RPC standards across clients. E.g if some clients respond with a hex string for block numbers and others respond with a number, that's going to make integration and adoption more challenging. The goal here is to be able assess EVM clients quickly and understand if they are implement in a way that conforms to a reference implementation (geth)

The approach taken here is to have a Test object that can executed in isolation. Each test should be an independent atomic unit that can executed in any order or in parallel. Since the tests themselves are objects, they are programmable. They can be sorted, fuzzed run at high speed, etc. Over time we'll add more edge cases, fuzzing capabilities, and additional name spaces like `debug` or `bor` and `zkevm`.

### CODE REVIEW

File: cmd/rpcfuzz/rpcfuzz.go [hunk 1]
```diff
@@ -0,0 +1,1395 @@
...
+		Validator: ValidateRegexString(`^0x[[:xdigit:]]*$`),
+	})
+
+	// cast rpc --rpc-url localhost:8545 eth_protocolVersion
+	allTests = append(allTests, &RPCTestGeneric{
+		Name:      "RPCTestEthProtocolVersion",
+		IsError:   true,
+		Method:    "eth_protocolVersion",
+		Args:      []interface{}{},
+		Validator: ValidateError(`method eth_protocolVersion does not exist`),
```
Comment (IdrisHanafi) [inline cmd/rpcfuzz/rpcfuzz.go:191]: nit and maybe something for me to pickup. regex matches might be risky here as there doesn't seem to be a clear consensus on what error message to return (even in the spec [doc](https://www.jsonrpc.org/specification) examples showing different error message). safer way would be to verify for a JSON RPC status code of -32601.

File: cmd/rpcfuzz/rpcfuzz.go [hunk 2]
```diff
@@ -0,0 +1,1395 @@
...
+		Method:    "eth_getStorageAt",
+		Args:      []interface{}{*testContractAddress, "0x3", "pending"},
+		Validator: ValidateRegexString(`^0x536f6c6964697479206279204578616d706c6500000000000000000000000026$`),
+	})
+	allTests = append(allTests, &RPCTestGeneric{
+		Name:      "RPCTestEthGetStorageAtZero",
+		Method:    "eth_getStorageAt",
+		Args:      []interface{}{*testContractAddress, "0x3", "0x0"},
+		Validator: ValidateRegexString(`^0x0{64}`),
+	})
```
Comment (IdrisHanafi) [inline cmd/rpcfuzz/rpcfuzz.go:313]: 🔥

File: cmd/rpcfuzz/rpcfuzz.go [hunk 3]
```diff
@@ -0,0 +1,1395 @@
...
+		RequiresUnlock: true,
+	})
+
+	// cast rpc --rpc-url localhost:8545 eth_sendRawTransaction '{"from": "0xb9b1cf51a65b50f74ed8bcb258413c02cba2ec57", "to": "0x85dA99c8a7C2C95964c8EfD687E95E632Fc533D6", "data": "0x", "gas": "0x5208", "gasPrice": "0x1", "nonce": "0x1"}'
+	allTests = append(allTests, &RPCTestDynamicArgs{
+		Name:      "RPCTestEthSendRawTransaction",
+		Method:    "eth_sendRawTransaction",
+		Args:      ArgsSignTransaction(ctx, rpcClient, &RPCTestTransactionArgs{To: testEthAddress.String(), Value: "0x123", Gas: "0x5208", Data: "0x", MaxFeePerGas: defaultMaxFeePerGas, MaxPriorityFeePerGas: defaultMaxPriorityFeePerGas}),
+		Validator: ValidateRegexString(`^0x[[:xdigit:]]{64}$`),
+	})
```
Comment (IdrisHanafi) [inline cmd/rpcfuzz/rpcfuzz.go:482]: niceee! in our previous implementation, we validated for correct balance changes of gas * gasPrice + amount sent. would it make sense to add it here or to put that at an EVM lower level test?

**Review by IdrisHanafi — APPROVED** [review]
lgtm!

just minor comments, which i can def address

File: cmd/rpcfuzz/rpcfuzz.go (hunk 1 above)
Comment (praetoriansentry) [reply cmd/rpcfuzz/rpcfuzz.go:191]: Interesting... Yeah my guess is these will fail across different clients. As we start to test and understand the ways in which they vary, I think we'll be in a good spot to generalize.

File: cmd/rpcfuzz/rpcfuzz.go (hunk 3 above)
Comment (praetoriansentry) [reply cmd/rpcfuzz/rpcfuzz.go:482]: It's a good question. I think it can't hurt. We could have some state / account level validators and chain them together with a `RequireAll`.

**Comment by IdrisHanafi** [comment]
just a clarifying question, it looks like we are completely steering away from checking for the client and EVM logic. From our previous version of conformance tests we didn't fully separate from checking the client implementation logic, and we found some bugs utilizing that version of conformance for zkEVM. Here is the example that it found, it conformed but that logic wasn't implemented correctly: https://github.com/0xPolygonHermez/zkevm-node/issues/1857
